src/Analyzers/AnalyzerFormatter.cs

```diff
@@ -130,40 +132,121 @@ static void LogDiagnosticLocations(Solution solution, IEnumerable<Diagnostic> di
 
         private async Task<Solution> FixDiagnosticsAsync(
             Solution solution,
-            ImmutableArray<(DiagnosticAnalyzer Analyzer, CodeFixProvider? Fixer)> analyzersAndFixers,
+            ImmutableArray<DiagnosticAnalyzer> allAnalyzers,
+            ImmutableArray<CodeFixProvider> allCodefixes,
             ImmutableDictionary<ProjectId, ImmutableHashSet<string>> projectDiagnostics,
             ImmutableHashSet<string> formattablePaths,
+            DiagnosticSeverity severity,
             ILogger logger,
             CancellationToken cancellationToken)
         {
-            // we need to run each codefix iteratively so ensure that all diagnostics are found and fixed
-            foreach (var (analyzer, codefix) in analyzersAndFixers)
+            // Determine the reported diagnostic ids
+            var reportedDiagnostics = projectDiagnostics.SelectMany(kvp => kvp.Value).Distinct().ToImmutableArray();
+            if (reportedDiagnostics.IsEmpty)
+            {
+                return solution;
+            }
+
+            // Build maps between diagnostic id and the associated analyzers and codefixes
+            var analyzersById = CreateAnalyzerMap(reportedDiagnostics, allAnalyzers);
+            var fixersById = CreateFixerMap(reportedDiagnostics, allCodefixes);
+
+            // We need to run each codefix iteratively so ensure that all diagnostics are found and fixed.
+            foreach (var diagnosticId in reportedDiagnostics)
             {
+                var analyzers = analyzersById[diagnosticId];
```

Comment:
perhaps we should try to get the value from the dictionary? Isn't it expected that we would not find entries for this?

Reply:
Ah I see, the `ToImmutableArray` on an empty enumerable will mean that we should always have an empty immutable array

---
In reply to: [450373591](https://github.com/dotnet/format/pull/723#discussion_r450373591) [](ancestors = 450373591)